planner/core/push_down_predicate.go

```diff
@@ -0,0 +1,249 @@
+	})
+
+	return exprGroups
+}
+
+/**
+ * @brief: This function is used to check if the condition contain heavy cost functions.
+ * @param: cond: condition of PhysicalSelection to be checked
+ * @note: heavy cost functions are functions that may cause a lot of memory allocation or disk IO.
+ * 		  For example, JSON functions.(To be supplemented)
+ */
+func withHeavyCostFunction(cond expression.Expression) bool {
+	if binop, ok := cond.(*expression.ScalarFunction); ok {
+		switch binop.FuncName.L {
+		// JSON functions
+		case ast.JSONType, ast.JSONExtract, ast.JSONUnquote, ast.JSONValid, ast.JSONSearch, ast.JSONContains, ast.JSONContainsPath, ast.JSONKeys, ast.JSONLength, ast.JSONDepth, ast.JSONQuote, ast.JSONSet, ast.JSONInsert, ast.JSONReplace, ast.JSONRemove, ast.JSONMerge, ast.JSONArrayAppend, ast.JSONArrayInsert, ast.JSONMergePatch, ast.JSONMergePreserve, ast.JSONPretty, ast.JSONStorageSize:
+			return true
+			// TODO: add more heavy cost functions
+		}
+	}
+	return false
+}
+
+/**
+ * @brief: This function is used to remove the conditions that needed to be pushed down.
+ * @param: physicalSelection: the PhysicalSelection to be modified
+ * @param: exprs: the conditions to be removed
+ */
+func removeSpecificExprsFromSelection(physicalSelection *PhysicalSelection, exprs []expression.Expression) {
+	newConds := make([]expression.Expression, 0, len(physicalSelection.Conditions))
+	for _, cond := range physicalSelection.Conditions {
+		if !expression.Contains(exprs, cond) {
+			newConds = append(newConds, cond)
+		}
+	}
+	physicalSelection.Conditions = newConds
+}
+
+/**
+ * @brief: This function is used to push down the some filter conditions of the selection to the tablescan.
+ * @param: sctx: the session context
+ * @param: physicalSelection: the PhysicalSelection containing the conditions to be pushed down
+ * @param: physicalTableScan: the PhysicalTableScan to be pushed down to
+ */
+func predicatePushDownToTableScanImpl(sctx sessionctx.Context, physicalSelection *PhysicalSelection, physicalTableScan *PhysicalTableScan) {
+	conds := physicalSelection.Conditions
+	if len(conds) == 0 {
+		return
+	}
+
+	// group the conditions by columns and sort them by selectivity
+	sortedConds := groupByColumnsSortBySelectivity(sctx, conds, physicalTableScan)
+
+	selectedConds := make([]expression.Expression, 0, len(conds))
+	selectedIncome := 0.0
+	selectedColumnCount := 0
+	totalColumnCount := len(physicalTableScan.Columns)
+
+	for _, exprGroup := range sortedConds {
+		mergedConds := append(selectedConds, exprGroup.exprs...)
+		selectivity, _, err := physicalTableScan.tblColHists.Selectivity(sctx, mergedConds, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		var cols []*expression.Column
+		cols = expression.ExtractColumnsFromExpressions(cols, mergedConds, nil)
```

Comment:
It seams that this function "ExtractColumnsFromExpressions" doesn't remove repeat column.
If the condition is "t1.a+t1.a = 1" , whether the size of cols is two ?

Reply:
fixed, use a map to remove duplicate columns.